File: app/scripts/utils/split-state.js
import { formatSecs } from './format.js';
import { parseTime } from './time.js';

const DEFAULT_VIEW = 'musician';
const SUPPORTED_VIEWS = ['musician', 'project'];

const formatMusicDuration = (seconds) => {
  if (seconds >= 3600) return formatSecs(seconds);

  const safeSeconds = Math.max(0, Math.round(seconds || 0));
  const minutes = Math.floor(safeSeconds / 60);
  const remainingSeconds = safeSeconds % 60;
  return `${String(minutes).padStart(2, '0')}:${String(remainingSeconds).padStart(2, '0')}`;
};

const normalizeSectionIndex = (value) => {
  if (Number.isInteger(value)) return value;
  const parsed = Number.parseInt(value, 10);
  return Number.isNaN(parsed) ? 0 : parsed;
};

const createSplitState = (seed = {}) => ({
  active: seed.active !== false,
  splitFromId: seed.splitFromId || null,
  splitTag: seed.splitTag || '',
  musicDuration: seed.musicDuration || '',
  estDuration: seed.estDuration || '',
  sectionIndex: normalizeSectionIndex(seed.sectionIndex),
});

export const getItemSplitParentIds = (item) => {
  const parentIds = new Set();

  if (item?.splitFromId) parentIds.add(item.splitFromId);
  if (item?.splitViews) {
    SUPPORTED_VIEWS.forEach((viewType) => {
      const parentId = item.splitViews?.[viewType]?.splitFromId;
      if (parentId) parentIds.add(parentId);
    });
  }

  return [...parentIds];
};

export const normalizeSplitViewType = (viewType) => (
  SUPPORTED_VIEWS.includes(viewType) ? viewType : DEFAULT_VIEW
);

export const createHiddenSplitState = () => ({
  active: false,
  splitFromId: null,
  splitTag: '',
  musicDuration: '',
  estDuration: '',
  sectionIndex: 0,
});

export const ensureItemSplitViews = (item) => {
  if (!item.splitViews) item.splitViews = {};

  const legacySeed = createSplitState({
    active: true,
    splitFromId: item.splitFromId,
    splitTag: item.splitTag,
    musicDuration: item.musicDuration,
    estDuration: item.estDuration,
    sectionIndex: item.sectionIndex,
  });

  SUPPORTED_VIEWS.forEach((viewType) => {
    if (item.splitViews[viewType]) {
      item.splitViews[viewType] = createSplitState(item.splitViews[viewType]);
    } else {
      item.splitViews[viewType] = { ...legacySeed };
    }
  });

  return item;
};

export const getItemSplitState = (item, viewType) => {
  ensureItemSplitViews(item);
  return item.splitViews[normalizeSplitViewType(viewType)];
};

export const peekItemSplitState = (item, viewType) => {
  const normalizedView = normalizeSplitViewType(viewType);
  const existingState = item?.splitViews?.[normalizedView];

  if (existingState) {
    return createSplitState(existingState);
  }

  return createSplitState({
    active: true,
    splitFromId: item?.splitFromId,
    splitTag: item?.splitTag,
    musicDuration: item?.musicDuration,
    estDuration: item?.estDuration,
    sectionIndex: item?.sectionIndex,
  });
};

export const setItemSplitState = (item, viewType, patch = {}) => {
  const current = getItemSplitState(item, viewType);
  item.splitViews[normalizeSplitViewType(viewType)] = createSplitState({
    ...current,
    ...patch,
  });
  return item.splitViews[normalizeSplitViewType(viewType)];
};

export const syncVisibleSplitDuration = (item, musicDuration, estDuration) => {
  ensureItemSplitViews(item);

  SUPPORTED_VIEWS.forEach((viewType) => {
    if (item.splitViews[viewType].active !== false) {
      item.splitViews[viewType] = createSplitState({
        ...item.splitViews[viewType],
        musicDuration,
        estDuration,
      });
    }
  });

  return item;
};

export const getVisibleConnectedSplitItems = (items, seedId, viewType) => {
  const connectedIds = getConnectedSplitItemIds(items, seedId);
  return items.filter((item) => (
    connectedIds.has(item.id) && peekItemVisibilityInView(item, viewType)
  ));
};

const getOrderedVisibleConnectedSplitItems = (items, seedId, viewType) => (
  getVisibleConnectedSplitItems(items, seedId, viewType).sort((a, b) => {
    const aState = peekItemSplitState(a, viewType);
    const bState = peekItemSplitState(b, viewType);

    if (aState.sectionIndex !== bState.sectionIndex) {
      return aState.sectionIndex - bState.sectionIndex;
    }

    return a.id.localeCompare(b.id);
  })
);

export const syncLegacySplitFields = (item, viewType) => {
  const state = getItemSplitState(item, viewType);

  item.splitFromId = state.splitFromId;
  item.musicDuration = state.musicDuration;
  item.estDuration = state.estDuration;
  item.sectionIndex = state.sectionIndex;

  if (state.splitTag) item.splitTag = state.splitTag;
  else delete item.splitTag;

  return item;
};

export const isItemVisibleInView = (item, viewType) => (
  getItemSplitState(item, viewType).active !== false
);

export const peekItemVisibilityInView = (item, viewType) => (
  peekItemSplitState(item, viewType).active !== false
);

export const deactivateItemInView = (item, viewType) => (
  setItemSplitState(item, viewType, createHiddenSplitState())
);

export const hasVisibleSplitStateInAnyView = (item) => {
  ensureItemSplitViews(item);
  return SUPPORTED_VIEWS.some((viewType) => item.splitViews[viewType].active !== false);
};

export const getConnectedSplitItemIds = (items, seedId) => {
  const connectedIds = new Set();
  const queue = [seedId];

  while (queue.length > 0) {
    const currentId = queue.shift();
    if (!currentId || connectedIds.has(currentId)) continue;

    connectedIds.add(currentId);

    items.forEach((item) => {
      const parentIds = getItemSplitParentIds(item);
      if (item.id === currentId || parentIds.includes(currentId)) {
        if (!connectedIds.has(item.id)) queue.push(item.id);
        parentIds.forEach((parentId) => {
          if (!connectedIds.has(parentId)) queue.push(parentId);
        });
      }
    });
  }

  return connectedIds;
};

export const rebalanceSplitFamilyDuration = (
  items,
  seedId,
  viewType,
  editedItemId,
  nextMusicDuration,
  estimateFn
) => {
  const visibleItems = getOrderedVisibleConnectedSplitItems(items, seedId, viewType);
  if (visibleItems.length < 2) {
    return { ok: true, totalMusicDuration: nextMusicDuration };
  }

  const totalSeconds = visibleItems.reduce(
    (sum, item) => sum + parseTime(peekItemSplitState(item, viewType).musicDuration || '00:00'),
    0
  );
  const nextSeconds = parseTime(nextMusicDuration || '00:00');
  const totalMusicDuration = formatMusicDuration(totalSeconds);

  if (nextSeconds > totalSeconds) {
    return {
      ok: false,
      reason: 'exceeds_total',
      totalMusicDuration,
    };
  }

  const targetItem = [...visibleItems].reverse().find((item) => item.id !== editedItemId);
  if (!targetItem) {
    return { ok: true, totalMusicDuration };
  }

  const fixedItems = visibleItems.filter((item) => (
    item.id !== editedItemId && item.id !== targetItem.id
  ));
  const fixedSeconds = fixedItems.reduce(
    (sum, item) => sum + parseTime(peekItemSplitState(item, viewType).musicDuration || '00:00'),
    0
  );
  const remainingSeconds = totalSeconds - nextSeconds - fixedSeconds;

  if (remainingSeconds < 0) {
    return {
      ok: false,
      reason: 'exceeds_total',
      totalMusicDuration,
    };
  }

  const editedItem = visibleItems.find((item) => item.id === editedItemId);
  if (editedItem) {
    const nextEst = typeof estimateFn === 'function'
      ? estimateFn(editedItem, nextMusicDuration)
      : peekItemSplitState(editedItem, viewType).estDuration;

    setItemSplitState(editedItem, viewType, {
      musicDuration: nextMusicDuration,
      estDuration: nextEst,
    });
  }

  const remainingMusicDuration = formatMusicDuration(remainingSeconds);
  const targetEst = typeof estimateFn === 'function'
    ? estimateFn(targetItem, remainingMusicDuration)
    : peekItemSplitState(targetItem, viewType).estDuration;

  setItemSplitState(targetItem, viewType, {
    musicDuration: remainingMusicDuration,
    estDuration: targetEst,
  });

  return {
    ok: true,
    totalMusicDuration,
    adjustedItemId: targetItem.id,
  };
};

export const syncFamilyTotalDuration = (items, seedId, sourceViewType, estimateFn) => {
  const connectedIds = getConnectedSplitItemIds(items, seedId);
  const sourceItems = getVisibleConnectedSplitItems(items, seedId, sourceViewType);

  const totalSeconds = sourceItems.reduce(
    (sum, item) => sum + parseTime(peekItemSplitState(item, sourceViewType).musicDuration || '00:00'),
    0
  );
  const totalMusicDuration = formatMusicDuration(totalSeconds);

  items.forEach((item) => {
    if (connectedIds.has(item.id)) {
      item.sharedMusicDuration = totalMusicDuration;
    }
  });

  SUPPORTED_VIEWS.forEach((viewType) => {
    const visibleItems = getVisibleConnectedSplitItems(items, seedId, viewType);
    if (visibleItems.length !== 1) return;

    const target = visibleItems[0];
    const currentState = getItemSplitState(target, viewType);
    const nextEst = typeof estimateFn === 'function'
      ? estimateFn(target, totalMusicDuration)
      : currentState.estDuration;

    setItemSplitState(target, viewType, {
      musicDuration: totalMusicDuration,
      estDuration: nextEst,
    });
  });

  return totalMusicDuration;
};

File: ios/MuscheCore/Sources/MuscheCore/SplitState.swift
import Foundation

/// `app/scripts/utils/split-state.js` 的直译移植（值类型 + 纯函数）。
/// 一条曲目在不同分类视图下可以各自拆分，结果存在 `splitViews[viewType]`，
/// 同时把当前视图的值同步回顶层旧字段（musicDuration / estDuration / splitTag）。
public enum SplitState {

    public static let defaultView = "musician"
    public static let supportedViews = ["musician", "project"]

    // MARK: - 内部

    private static func formatMusicDuration(_ seconds: Int) -> String {
        if seconds >= 3600 { return Format.formatSecs(seconds) }
        let safe = max(0, seconds)
        return String(format: "%02d:%02d", safe / 60, safe % 60)
    }

    private static func normalizeSectionIndex(_ value: Any?) -> Int {
        if let i = value as? Int { return i }
        if let d = value as? Double { return Int(d) }
        if let s = value as? String, let i = Int(s) { return i }
        return 0
    }

    /// 复刻 JS `createSplitState(seed)`：应用 `||` 归一化。
    private static func makeState(
        active: Bool? = nil,
        splitFromId: String? = nil,
        splitTag: String? = nil,
        musicDuration: String? = nil,
        estDuration: String? = nil,
        sectionIndex: Any? = nil
    ) -> SplitViewState {
        SplitViewState(
            active: active ?? true,
            splitFromId: (splitFromId?.isEmpty == false) ? splitFromId : nil,
            splitTag: splitTag ?? "",
            musicDuration: musicDuration ?? "",
            estDuration: estDuration ?? "",
            sectionIndex: normalizeSectionIndex(sectionIndex ?? 0)
        )
    }

    // MARK: - 导出函数

    /// 返回该条目在所有视图里的父 id 集合。
    public static func getItemSplitParentIds(_ item: PoolItem) -> [String] {
        var parents = Set<String>()
        if let pid = item.splitFromId, !pid.isEmpty { parents.insert(pid) }
        if let views = item.splitViews {
            for viewType in supportedViews {
                if let pid = views[viewType]?.splitFromId, !pid.isEmpty { parents.insert(pid) }
            }
        }
        return Array(parents)
    }

    public static func normalizeSplitViewType(_ viewType: String) -> String {
        supportedViews.contains(viewType) ? viewType : defaultView
    }

    public static func createHiddenSplitState() -> SplitViewState {
        SplitViewState(active: false, splitFromId: nil, splitTag: "", musicDuration: "", estDuration: "", sectionIndex: 0)
    }

    /// 保证 splitViews 存在并为两个视图各生成一个拆分状态（缺失的用顶层旧字段做种子）。
    @discardableResult
    public static func ensureItemSplitViews(_ item: inout PoolItem) -> PoolItem {
        if item.splitViews == nil { item.splitViews = SplitViews() }

        let legacySeed = makeState(
            active: true,
            splitFromId: item.splitFromId,
            splitTag: item.splitTag,
            musicDuration: item.musicDuration,
            estDuration: item.estDuration,
            sectionIndex: item.sectionIndex
        )

        for viewType in supportedViews {
            if item.splitViews![viewType] == nil {
                item.splitViews![viewType] = legacySeed
            }
        }
        return item
    }

    /// 取某视图的拆分状态（可能因 ensure 产生副作用）。
    public static func getItemSplitState(_ item: inout PoolItem, _ viewType: String) -> SplitViewState {
        ensureItemSplitViews(&item)
        return item.splitViews![normalizeSplitViewType(viewType)]!
    }

    /// 只读地取某视图拆分状态（不修改条目；缺省时用顶层旧字段构造临时态）。
    public static func peekItemSplitState(_ item: PoolItem, _ viewType: String) -> SplitViewState {
        let normalized = normalizeSplitViewType(viewType)
        if let existing = item.splitViews?[normalized] {
            return existing
        }
        return makeState(
            active: true,
            splitFromId: item.splitFromId,
            splitTag: item.splitTag,
            musicDuration: item.musicDuration,
            estDuration: item.estDuration,
            sectionIndex: item.sectionIndex
        )
    }

    /// 合并 patch 到某视图的拆分状态。
    @discardableResult
    public static func setItemSplitState(_ item: inout PoolItem, _ viewType: String, _ patch: (inout SplitViewState) -> Void) -> SplitViewState {
        var current = getItemSplitState(&item, viewType)
        patch(&current)
        current = makeState(
            active: current.active,
            splitFromId: current.splitFromId,
            splitTag: current.splitTag,
            musicDuration: current.musicDuration,
            estDuration: current.estDuration,
            sectionIndex: current.sectionIndex
        )
        item.splitViews![normalizeSplitViewType(viewType)] = current
        return current
    }

    /// 把时长同步到所有「未隐藏」视图的拆分状态。
    @discardableResult
    public static func syncVisibleSplitDuration(_ item: inout PoolItem, _ musicDuration: String, _ estDuration: String) -> PoolItem {
        ensureItemSplitViews(&item)
        for viewType in supportedViews {
            if item.splitViews![viewType]!.active != false {
                item.splitViews![viewType] = makeState(
                    active: item.splitViews![viewType]!.active,
                    splitFromId: item.splitViews![viewType]!.splitFromId,
                    splitTag: item.splitViews![viewType]!.splitTag,
                    musicDuration: musicDuration,
                    estDuration: estDuration,
                    sectionIndex: item.splitViews![viewType]!.sectionIndex
                )
            }
        }
        return item
    }

    /// 取某视图下「连通且可见」的拆分条目。
    public static func getVisibleConnectedSplitItems(_ items: [PoolItem], _ seedId: String, _ viewType: String) -> [PoolItem] {
        let connected = getConnectedSplitItemIds(items, seedId)
        return items.filter { connected.contains($0.id) && peekItemVisibilityInView($0, viewType) }
    }

    /// 把当前视图的拆分状态同步回顶层旧字段。
    @discardableResult
    public static func syncLegacySplitFields(_ item: inout PoolItem, _ viewType: String) -> PoolItem {
        let state = getItemSplitState(&item, viewType)
        item.splitFromId = state.splitFromId
        item.musicDuration = state.musicDuration
        item.estDuration = state.estDuration
        item.sectionIndex = state.sectionIndex
        if !state.splitTag.isEmpty {
            item.splitTag = state.splitTag
        } else {
            item.splitTag = nil
        }
        return item
    }

    /// 某条目在某视图是否可见（active !== false）。
    public static func isItemVisibleInView(_ item: inout PoolItem, _ viewType: String) -> Bool {
        getItemSplitState(&item, viewType).active != false
    }

    /// 只读地判可见性。
    public static func peekItemVisibilityInView(_ item: PoolItem, _ viewType: String) -> Bool {
        peekItemSplitState(item, viewType).active != false
    }

    /// 把某视图的拆分状态置为隐藏态。
    @discardableResult
    public static func deactivateItemInView(_ item: inout PoolItem, _ viewType: String) -> SplitViewState {
        setItemSplitState(&item, viewType) { $0 = createHiddenSplitState() }
    }

    /// 任一视图仍有可见拆分状态。
    public static func hasVisibleSplitStateInAnyView(_ item: inout PoolItem) -> Bool {
        ensureItemSplitViews(&item)
        return supportedViews.contains { item.splitViews![$0]!.active != false }
    }

    /// 广度遍历，返回与 seedId 连通的所有条目 id（含 seed 自己与父链）。
    public static func getConnectedSplitItemIds(_ items: [PoolItem], _ seedId: String) -> Set<String> {
        var connected = Set<String>()
        var queue = [seedId]

        while !queue.isEmpty {
            let currentId = queue.removeFirst()
            if currentId.isEmpty || connected.contains(currentId) { continue }
            connected.insert(currentId)

            for item in items {
                let parentIds = getItemSplitParentIds(item)
                if item.id == currentId || parentIds.contains(currentId) {
                    if !connected.contains(item.id) { queue.append(item.id) }
                    for pid in parentIds where !connected.contains(pid) { queue.append(pid) }
                }
            }
        }
        return connected
    }
}
